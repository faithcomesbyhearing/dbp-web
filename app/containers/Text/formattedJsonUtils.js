/**
 * utils.js
 *
 * Utility functions for FormattedText component.
 */

/**
 * Checks if a verse number falls within a given range (start/end).
 */
export const isVerseInRange = (start, targetVerse, end) => {
	// Ensure inputs are valid numbers before comparison
	if (typeof start !== 'number' || typeof targetVerse !== 'number') return false;
	// If end is not a number or null/undefined, treat the range as a single verse (start)
	const endVerse = typeof end === 'number' ? end : start;
	return targetVerse >= start && targetVerse <= endVerse;
};

/**
 * Calculates display properties for a verse based on annotations and player state.
 */
export const calculateVerseProps = (
	verseNumber,
	currentPlayingVerse,
	currentSelectedVerse, // Can be number or { verseStart, verseEnd }
	annotations,
) => {
	if (typeof verseNumber !== 'number') {
		// Not within a verse scope
		return {
			isSelectable: false,
			isPlaying: false,
			isSelected: false,
			hasNote: false,
			hasBookmark: false,
			highlightData: null,
		};
	}

	const isPlaying = currentPlayingVerse === verseNumber;

	// Handle different shapes of currentSelectedVerse
	const selectionStart =
		currentSelectedVerse?.verseStart ?? currentSelectedVerse?.verse ?? null;
	const selectionEnd = currentSelectedVerse?.verseEnd ?? null; // Use null if no end provided
	const isSelected = isVerseInRange(
		selectionStart,
		verseNumber,
		selectionEnd, // Pass null if single verse selection
	);

	const notes = annotations?.notes || [];
	const bookmarks = annotations?.bookmarks || [];
	const highlights = annotations?.highlights || [];

	// Use find with isVerseInRange for consistency
	const note = notes.find((n) =>
		isVerseInRange(n.verse_start, verseNumber, n.verse_end),
	);
	const bookmark = bookmarks.find((b) => b.verse === verseNumber); // Bookmarks often single verse
	const highlight = highlights.find((hl) =>
		isVerseInRange(hl.verse_start, verseNumber, hl.verse_end),
	);

	return {
		isSelectable: true, // Verses are selectable
		isPlaying,
		isSelected,
		hasNote: !!note,
		hasBookmark: !!bookmark,
		highlightData: highlight || null, // Pass highlight object or null
	};
};

// Helper to find a specific subtype recursively using array methods
export const findSubtypeRecursively = (content, subtypeToFind) => {
	if (!Array.isArray(content)) return null;

	// Find directly in the current array
	const directMatch = content.find(
		(item) =>
			typeof item === 'object' &&
			item !== null &&
			item.subtype === subtypeToFind,
	);
	if (directMatch) {
		return directMatch;
	}

	// If not found directly, search recursively in children
	// Use reduce to find the first match in any child's content
	return content.reduce((found, item) => {
		if (found) return found; // Already found in a previous item's children
		if (typeof item === 'object' && item !== null && item.content) {
			return findSubtypeRecursively(item.content, subtypeToFind);
		}
		return null; // Not found in this item's children
	}, null);
};

// Extracts text using array methods
export function extractTextRecursive(contentArray) {
	if (!Array.isArray(contentArray)) return '';

	const text = contentArray.reduce((acc, item) => {
		if (typeof item === 'string') {
			return acc + item;
		} else if (typeof item === 'object' && item !== null) {
			// Recurse into content, avoid adding marker text like note_caller itself
			let childText = '';
			if (item.content && item.subtype !== 'note_caller') {
				childText = extractTextRecursive(item.content);
			} else if (
				item.sequence?.blocks?.[0]?.content &&
				item.subtype !== 'note_caller'
			) {
				// Handle grafts (excluding note_caller itself)
				childText = extractTextRecursive(item.sequence.blocks[0].content);
			}
			return acc + childText;
		}
		return acc; // Ignore other types
	}, '');

	// Clean up spacing after accumulating all text
	return text.replace(/\s+/g, ' ').trim();
}

/**
 * Traverses content arrays to find footnote grafts and extract data.
 * Uses array methods and returns the updated index.
 */
function findAndExtractFootnotesRecursiveWithIndex(
	contentArray,
	footnotesMap, // The map to add footnotes to
	startIndex, // The starting index for this level
) {
	if (!Array.isArray(contentArray)) return startIndex; // Return index unchanged

	let currentIndex = startIndex; // Use a local variable to track index

	contentArray.forEach((item) => {
		if (typeof item === 'object' && item !== null) {
			// Check if this item IS a footnote graft
			if (
				item.type === 'graft' &&
				item.subtype === 'footnote' &&
				item.sequence?.blocks?.[0]?.content
			) {
				const footnoteBlockContent = item.sequence.blocks[0].content;
				let marker = '?';
				let footnoteText = '';

				try {
					const callerGraft = footnoteBlockContent.find(
						(c) => c?.type === 'graft' && c?.subtype === 'note_caller',
					);
					marker = callerGraft?.sequence?.blocks?.[0]?.content?.[0] || marker;

					const textWrapper = footnoteBlockContent.find(
						(c) => c?.type === 'wrapper' && c?.subtype === 'usfm:ft',
					);
					if (textWrapper?.content) {
						footnoteText = extractTextRecursive(textWrapper.content);
					}
				} catch (e) {
					// Keep error log for parsing issues
					console.error('Error parsing footnote graft:', e, item); // eslint-disable-line no-console
				}

				// Generate key using a unique identifier (text-based is more stable than index)
				const generatedKey = `fn-${
					footnoteText.replace(/\s+/g, '_') || currentIndex
				}`;
				// Add data to the map object (passed by reference)
				// No no-param-reassign error here as we modify the object's properties, not the reference itself
				// eslint-disable-next-line no-param-reassign
				footnotesMap[generatedKey] = { text: footnoteText, caller: marker };
				// Increment local index tracker
				currentIndex += 1; // Replaced ++
			}

			// Standard recursion into 'content', passing the current index
			if (item.content) {
				currentIndex = findAndExtractFootnotesRecursiveWithIndex(
					item.content,
					footnotesMap,
					currentIndex, // Pass the potentially updated index
				);
			}
			// Standard recursion into 'sequence.blocks', passing the current index
			if (item.sequence?.blocks) {
				currentIndex = findAndExtractFootnotesFromBlocksWithIndex(
					item.sequence.blocks,
					footnotesMap,
					currentIndex, // Pass the potentially updated index
				);
			}
		}
	});
	return currentIndex; // Return the final index after processing this array
}

/**
 * Helper to iterate over blocks using array methods and start recursive search.
 */
function findAndExtractFootnotesFromBlocksWithIndex(
	blocks,
	footnotesMap, // Pass the map
	startIndex, // Pass the starting index
) {
	if (!Array.isArray(blocks)) return startIndex; // Return index unchanged

	// Use reduce to iterate and track the index accumulator
	return blocks.reduce((currentIndexAcc, block) => {
		let updatedIndex = currentIndexAcc; // Start with the index from previous block/initial call
		if (block.content) {
			updatedIndex = findAndExtractFootnotesRecursiveWithIndex(
				block.content,
				footnotesMap,
				updatedIndex, // Pass current index to recursive call
			);
		}
		if (block.type === 'graft' && block.sequence?.blocks) {
			// Recurse into graft blocks
			updatedIndex = findAndExtractFootnotesFromBlocksWithIndex(
				block.sequence.blocks,
				footnotesMap,
				updatedIndex, // Pass current index to recursive call
			);
		}
		return updatedIndex; // Return the index after processing this block
	}, startIndex); // Initial value for the accumulator is startIndex
}

/**
 * Main function to extract footnotes into a map.
 * Uses text-based keys for better stability.
 */
export function extractFootnotesWithIndexKey(chapterJson) {
	const footnotesMap = {};
	if (!chapterJson?.sequence?.blocks) {
		return footnotesMap;
	}
	// Start traversal from the main sequence blocks with initial index 0
	findAndExtractFootnotesFromBlocksWithIndex(
		chapterJson.sequence.blocks,
		footnotesMap, // Pass the map to be populated
		0, // Initial index (less critical now with text-based keys)
	);
	return footnotesMap;
}
