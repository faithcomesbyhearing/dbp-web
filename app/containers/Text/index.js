/**
 *
 * Text
 * TODO: Split this up into components and isolate the different parts that do not need to be together
 * TODO: Find way to highlight and un-highlight verses without directly manipulating the dom by adding/removing classnames
 * TODO: Paramaterize addHighlight so it can be more easily tested and can be re-used
 */

import React from 'react';
import PropTypes from 'prop-types';
import dynamic from 'next/dynamic';
import LoadingSpinner from '../../components/LoadingSpinner';
import getPreviousChapterUrl from '../../utils/getPreviousChapterUrl';
import getNextChapterUrl from '../../utils/getNextChapterUrl';
import {
	getClassNameForTextContainer,
	isEndOfBible,
	isStartOfBible,
} from './textRenderUtils';
import NewChapterArrow from '../../components/NewChapterArrow';
const Verses = dynamic(import('../Verses'), {
	loading: () => <LoadingSpinner />,
});

/* Disabling the jsx-a11y linting because we need to capture the selected text
	 and the most straight forward way of doing so is with the onMouseUp event */
// Todo: Set selected text when user clicks a verse
class Text extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			loadingNextPage: false,
		};
	}

	componentDidMount() {
		// Doing all these assignments because nextjs was erroring because they try to use the dom
		if (this.mainWrapper) {
			this.mainWrapper.focus();
		}
	}

	componentDidUpdate(prevProps) {
		// If there is new formatted text or new plain text then the menus need to be disabled
		// Change the loading state to be set and controlled within the API call and promise
		if (
			// If there was a change in the text at all then the menus need to be closed
			prevProps.verseNumber !== this.props.verseNumber ||
			prevProps.activeChapter !== this.props.activeChapter ||
			prevProps.activeBookId !== this.props.activeBookId ||
			prevProps.activeTextId !== this.props.activeTextId
		) {
			this.setState({ loadingNextPage: false });
		}
	}

	handleArrowClick = () => {
		this.setState({ loadingNextPage: true });
	};

	mainWrapperRef = (el) => {
		this.mainWrapper = el;
	};

	render() {
		const {
			activeChapter,
			text,
			loadingNewChapterText,
			verseNumber,
			activeTextId,
			activeBookId,
			books,
			menuIsOpen,
			isScrollingDown,
			audioPlayerState,
			subFooterOpen,
			chapterTextLoadingState,
			videoPlayerOpen,
			hasVideo,
			audioType,
			changingVersion,
		} = this.props;
		const { loadingNextPage } = this.state;

		if (
			loadingNewChapterText ||
			chapterTextLoadingState ||
			loadingNextPage ||
			changingVersion
		) {
			return (
				<div
					className={getClassNameForTextContainer({
						isScrollingDown,
						videoPlayerOpen,
						subFooterOpen,
						hasVideo,
					})}
				>
					<LoadingSpinner />
				</div>
			);
		}

		return (
			<div
				id="text-container-parent"
				className={getClassNameForTextContainer({
					isScrollingDown,
					videoPlayerOpen,
					subFooterOpen,
					hasVideo,
					audioPlayerState,
				})}
			>
				<NewChapterArrow
					getNewUrl={getPreviousChapterUrl}
					urlProps={{
						books,
						chapter: activeChapter,
						bookId: activeBookId.toLowerCase(),
						textId: activeTextId.toLowerCase(),
						verseNumber,
						text,
						audioType,
					}}
					clickHandler={this.handleArrowClick}
					disabled={
						isStartOfBible(books, activeBookId, activeChapter) || menuIsOpen
					}
					svgid={'arrow_left'}
					svgClasses={'prev-arrow-svg'}
					disabledContainerClasses={'arrow-wrapper prev disabled'}
					containerClasses={'arrow-wrapper prev'}
				/>
				<div ref={this.mainWrapperRef} className={'main-wrapper main-text-wrapper'}>
					<Verses menuIsOpen={menuIsOpen} />
				</div>
				<NewChapterArrow
					getNewUrl={getNextChapterUrl}
					urlProps={{
						books,
						chapter: activeChapter,
						bookId: activeBookId.toLowerCase(),
						textId: activeTextId.toLowerCase(),
						verseNumber,
						text,
						audioType,
					}}
					clickHandler={this.handleArrowClick}
					disabled={
						isEndOfBible(books, activeBookId, activeChapter) || menuIsOpen
					}
					svgid={'arrow_right'}
					svgClasses={'next-arrow-svg'}
					disabledContainerClasses={'arrow-wrapper next disabled'}
					containerClasses={'arrow-wrapper next'}
				/>
			</div>
		);
	}
}

Text.propTypes = {
	text: PropTypes.array,
	books: PropTypes.array,
	activeChapter: PropTypes.number,
	hasVideo: PropTypes.bool,
	menuIsOpen: PropTypes.bool,
	subFooterOpen: PropTypes.bool,
	videoPlayerOpen: PropTypes.bool,
	changingVersion: PropTypes.bool,
	isScrollingDown: PropTypes.bool,
	audioPlayerState: PropTypes.bool,
	loadingNewChapterText: PropTypes.bool,
	chapterTextLoadingState: PropTypes.bool,
	audioType: PropTypes.string,
	verseNumber: PropTypes.string,
	activeTextId: PropTypes.string.isRequired,
	activeBookId: PropTypes.string,
};

export default Text;
