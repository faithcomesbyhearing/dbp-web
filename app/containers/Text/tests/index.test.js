import React from 'react';
import { render } from '@testing-library/react';
import Text from '../index';

// Basic Text Props
const text = [
	{
		book_id: 'LUK',
		book_name: 'Luke',
		book_name_alt: 'Luke',
		chapter: 1,
		chapter_alt: '1',
		verse_start: 1,
		verse_start_alt: '1',
		verse_end: 1,
		verse_end_alt: '1',
		verse_text:
			'Since many have undertaken to compile a narrative of the events that have been fulfilled among us,',
	},
	{
		book_id: 'LUK',
		book_name: 'Luke',
		book_name_alt: 'Luke',
		chapter: 1,
		chapter_alt: '1',
		verse_start: 2,
		verse_start_alt: '2',
		verse_end: 2,
		verse_end_alt: '2',
		verse_text:
			'just as those who were eyewitnesses from the beginning and ministers of the word have handed them down to us,',
	},
	{
		book_id: 'LUK',
		book_name: 'Luke',
		book_name_alt: 'Luke',
		chapter: 1,
		chapter_alt: '1',
		verse_start: 3,
		verse_start_alt: '3',
		verse_end: 3,
		verse_end_alt: '3',
		verse_text:
			'I too have decided, after investigating everything accurately anew, to write it down in an orderly sequence for you, most excellent Theophilus,',
	},
	{
		book_id: 'LUK',
		book_name: 'Luke',
		book_name_alt: 'Luke',
		chapter: 1,
		chapter_alt: '1',
		verse_start: 4,
		verse_start_alt: '4',
		verse_end: 4,
		verse_end_alt: '4',
		verse_text:
			'so that you may realize the certainty of the teachings you have received.',
	},
	{
		book_id: 'LUK',
		book_name: 'Luke',
		book_name_alt: 'Luke',
		chapter: 1,
		chapter_alt: '1',
		verse_start: 5,
		verse_start_alt: '5',
		verse_end: 5,
		verse_end_alt: '5',
		verse_text:
			'In the days of Herod, King of Judea, there was a priest named Zechariah of the priestly division of Abijah; his wife was from the daughters of Aaron, and her name was Elizabeth.',
	},
];
const books = [
	{
		book_id: 'GEN',
		book_id_usfx: 'GN',
		book_id_osis: 'Gen',
		name: 'Genesis',
		testament: 'OT',
		testament_order: 1,
		book_order: 1,
		book_group: 'The Law',
		chapters: [
			1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
			22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39,
			40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
		],
	},
	{
		book_id: 'EXO',
		book_id_usfx: 'EX',
		book_id_osis: 'Exod',
		name: 'Exodus',
		testament: 'OT',
		testament_order: 2,
		book_order: 2,
		book_group: 'The Law',
		chapters: [
			1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
			22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39,
			40,
		],
	},
	{
		book_id: 'LEV',
		book_id_usfx: 'LV',
		book_id_osis: 'Lev',
		name: 'Leviticus',
		testament: 'OT',
		testament_order: 3,
		book_order: 3,
		book_group: 'The Law',
		chapters: [
			1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
			22, 23, 24, 25, 26, 27,
		],
	},
];
const activeChapter = 1;
const hasVideo = true;
const menuIsOpen = true;
const subFooterOpen = true;
const videoPlayerOpen = true;
const changingVersion = true;
const isScrollingDown = true;
const audioPlayerState = true;
const loadingNewChapterText = true;
const chapterTextLoadingState = true;
const audioType = 'audio_drama';
const verseNumber = '';
const activeTextId = 'ENGESV';
const activeBookId = 'LEV';

describe('<Text />', () => {
	it('Should match previous snapshot with all options true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				hasVideo={hasVideo}
				menuIsOpen={menuIsOpen}
				subFooterOpen={subFooterOpen}
				videoPlayerOpen={videoPlayerOpen}
				changingVersion={changingVersion}
				isScrollingDown={isScrollingDown}
				audioPlayerState={audioPlayerState}
				loadingNewChapterText={loadingNewChapterText}
				chapterTextLoadingState={chapterTextLoadingState}
				audioType={audioType}
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with no options true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with hasVideo true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				hasVideo
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with menuIsOpen true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				menuIsOpen
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with subFooterOpen true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				subFooterOpen
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with videoPlayerOpen true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				videoPlayerOpen
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with changingVersion true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				changingVersion
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with audioPlayerState true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				audioPlayerState
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with loadingNewChapterText true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				loadingNewChapterText
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with chapterTextLoadingState true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				chapterTextLoadingState
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with loadingNewChapterText true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				loadingNewChapterText
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('Should match previous snapshot with chapterTextLoadingState true', () => {
		const { asFragment } = render(
			<Text
				text={text}
				books={books}
				activeChapter={activeChapter}
				audioType={audioType}
				chapterTextLoadingState
				verseNumber={verseNumber}
				activeTextId={activeTextId}
				activeBookId={activeBookId}
			/>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
