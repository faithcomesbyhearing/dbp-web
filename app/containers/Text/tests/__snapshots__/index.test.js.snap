// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> Should match previous snapshot with all options true 1`] = `
<div
  className="text-container scrolled-down video-player-open has-video"
>
  <div
    className="loading-spinner"
  >
    <span
      className="loading-spinner1 loading-spinner-child"
    />
    <span
      className="loading-spinner2 loading-spinner-child"
    />
    <span
      className="loading-spinner3 loading-spinner-child"
    />
    <span
      className="loading-spinner4 loading-spinner-child"
    />
    <span
      className="loading-spinner5 loading-spinner-child"
    />
    <span
      className="loading-spinner6 loading-spinner-child"
    />
    <span
      className="loading-spinner7 loading-spinner-child"
    />
    <span
      className="loading-spinner8 loading-spinner-child"
    />
    <span
      className="loading-spinner9 loading-spinner-child"
    />
    <span
      className="loading-spinner10 loading-spinner-child"
    />
    <span
      className="loading-spinner11 loading-spinner-child"
    />
    <span
      className="loading-spinner12 loading-spinner-child"
    />
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with audioPlayerState true 1`] = `
<div
  className="text-container audio-player-open"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="prev-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_left"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="next-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_right"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with changingVersion true 1`] = `
<div
  className="text-container"
>
  <div
    className="loading-spinner"
  >
    <span
      className="loading-spinner1 loading-spinner-child"
    />
    <span
      className="loading-spinner2 loading-spinner-child"
    />
    <span
      className="loading-spinner3 loading-spinner-child"
    />
    <span
      className="loading-spinner4 loading-spinner-child"
    />
    <span
      className="loading-spinner5 loading-spinner-child"
    />
    <span
      className="loading-spinner6 loading-spinner-child"
    />
    <span
      className="loading-spinner7 loading-spinner-child"
    />
    <span
      className="loading-spinner8 loading-spinner-child"
    />
    <span
      className="loading-spinner9 loading-spinner-child"
    />
    <span
      className="loading-spinner10 loading-spinner-child"
    />
    <span
      className="loading-spinner11 loading-spinner-child"
    />
    <span
      className="loading-spinner12 loading-spinner-child"
    />
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with chapterTextLoadingState true 1`] = `
<div
  className="text-container"
>
  <div
    className="loading-spinner"
  >
    <span
      className="loading-spinner1 loading-spinner-child"
    />
    <span
      className="loading-spinner2 loading-spinner-child"
    />
    <span
      className="loading-spinner3 loading-spinner-child"
    />
    <span
      className="loading-spinner4 loading-spinner-child"
    />
    <span
      className="loading-spinner5 loading-spinner-child"
    />
    <span
      className="loading-spinner6 loading-spinner-child"
    />
    <span
      className="loading-spinner7 loading-spinner-child"
    />
    <span
      className="loading-spinner8 loading-spinner-child"
    />
    <span
      className="loading-spinner9 loading-spinner-child"
    />
    <span
      className="loading-spinner10 loading-spinner-child"
    />
    <span
      className="loading-spinner11 loading-spinner-child"
    />
    <span
      className="loading-spinner12 loading-spinner-child"
    />
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with hasVideo true 1`] = `
<div
  className="text-container has-video"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="prev-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_left"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="next-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_right"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with loadingNewChapterText true 1`] = `
<div
  className="text-container"
>
  <div
    className="loading-spinner"
  >
    <span
      className="loading-spinner1 loading-spinner-child"
    />
    <span
      className="loading-spinner2 loading-spinner-child"
    />
    <span
      className="loading-spinner3 loading-spinner-child"
    />
    <span
      className="loading-spinner4 loading-spinner-child"
    />
    <span
      className="loading-spinner5 loading-spinner-child"
    />
    <span
      className="loading-spinner6 loading-spinner-child"
    />
    <span
      className="loading-spinner7 loading-spinner-child"
    />
    <span
      className="loading-spinner8 loading-spinner-child"
    />
    <span
      className="loading-spinner9 loading-spinner-child"
    />
    <span
      className="loading-spinner10 loading-spinner-child"
    />
    <span
      className="loading-spinner11 loading-spinner-child"
    />
    <span
      className="loading-spinner12 loading-spinner-child"
    />
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with menuIsOpen true 1`] = `
<div
  className="text-container"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev disabled"
  />
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next disabled"
  />
</div>
`;

exports[`<Text /> Should match previous snapshot with no options true 1`] = `
<div
  className="text-container"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="prev-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_left"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="next-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_right"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with subFooterOpen true 1`] = `
<div
  className="text-container sub-footer-open"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="prev-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_left"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="next-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_right"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
</div>
`;

exports[`<Text /> Should match previous snapshot with videoPlayerOpen true 1`] = `
<div
  className="text-container"
  id="text-container-parent"
>
  <div
    className="arrow-wrapper prev"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="prev-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_left"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <div
    className="main-wrapper main-text-wrapper"
  />
  <div
    className="arrow-wrapper next"
    id=""
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    title=""
  >
    <svg
      className="next-arrow-svg"
    >
      <use
        xlinkHref="/svglist.svg#arrow_right"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
</div>
`;
